Source/ROLib/Modules/ROLAnimateEngineHeat.cs

```diff
@@ -0,0 +1,195 @@
+        //will reach max glow at this temp, and begin dissipating even faster past this point
+        [KSPField]
+        public float maxHeat = 2400;
+
+        //maxStoredHeat
+        //storedHeat will not go beyond this, sets retention period for maximum glow
+        [KSPField]
+        public float maxStoredHeat = 3600;
+
+        //curve to adjust heat dissipation; should generally expel heat faster when hotter
+        [KSPField]
+        public FloatCurve heatDissipationCurve = new FloatCurve();
+
+        //the heat-output curve for an engine (varies with thrust/throttle), in case it is not linear
+        [KSPField]
+        public FloatCurve heatAccumulationCurve = new FloatCurve();
+
+        [KSPField]
+        public FloatCurve redCurve = new FloatCurve();
+
+        [KSPField]
+        public FloatCurve blueCurve = new FloatCurve();
+
+        [KSPField]
+        public FloatCurve greenCurve = new FloatCurve();
+
+        [KSPField]
+        public string engineID = "Engine";
+
+        [KSPField]
+        public String meshName = String.Empty;
+
+        [KSPField(isPersistant = true)]
+        public float currentHeat = 0;
+
+        [KSPField]
+        public bool useThrottle;
+
+        int shaderEmissiveID;
+
+        private ModuleEngines engineModule;
+
+        private Renderer[] animatedRenderers;
+
+        private Color emissiveColor = new Color(0f, 0f, 0f, 1f);
+
+        public override void OnAwake()
+        {
+            heatDissipationCurve.Add(0f, 0.2f);
+            heatDissipationCurve.Add(1f, 1f);
+
+            heatAccumulationCurve.Add(0f, 0f);
+            heatAccumulationCurve.Add(1f, 1f);
+            
+            redCurve.Add(0f, 0f);
+            redCurve.Add(1f, 1f);
+
+            blueCurve.Add(0f, 0f);
+            blueCurve.Add(1f, 1f);
+
+            greenCurve.Add(0f, 0f);
+            greenCurve.Add(1f, 1f);
+
+            shaderEmissiveID = Shader.PropertyToID("_EmissiveColor");
+        }
+
+        public override void OnStart(StartState state)
+        {
+            locateAnimatedTransforms();
+            locateEngineModule();
+        }
+
+        public void FixedUpdate()
+        {
+            if (HighLogic.LoadedSceneIsFlight)
+                updateHeat();
+        }
+
+        private void locateEngineModule()
+        {
+            engineModule = null;
+            ModuleEngines[] engines = part.GetComponents<ModuleEngines>();
+            int len = engines.Length;
+            engineModule = part.GetComponents<ModuleEngines>().FirstOrDefault(x => x.engineID == engineID);
+            if (engineModule == null)
+            {
+                MonoBehaviour.print("ERROR: Could not locate engine by ID: " + engineID + " for part: " + part + " for ROLAnimateEngineHeat.  This will cause errors during gameplay.  Setting engine to first engine module (if present)");
+                if (engines.Length > 0) { engineModule = engines[0]; }
+            }
```

Comment:
Suggest
```
engineModule = part.GetComponents<ModuleEngines>().FirstOrDefault(x => x.engineID == engineID);
if (engineModule == null) MonoBehaviour.print("blah...");
engineModule ??= part.GetComponents<ModuleEngines>().FirstOrDefault();
```
And that's it for the entire method.
First line tries to get the line you want.
Second line screams about a config error.
Third line uses a null coalescing operator.  It only assigns if the left-hand side (engineModule) is null.  And it just gets the first ModuleEngines if any exist.

Reply:
What would happen if there are no ModuleEngines? What is the default of ModuleEngines?

Reply:
If it is null, we could just use GetComponent<>. Also, we should probably not let our Module just throw NREs?

Reply:
You're right that GetComponent() is a better idea than GetComponents().FirstOrDefault().

If there are no ModuleEngines, then engineModule will be null at the end.  I believe this is what we want, and code accessing engineModule needs to be aware it can be null.  Right, throwing NREs is not a good idea.

Reply:
FirstOrDefault returns null for classes, so just saying `part.GetComponent<ModuleEngines>` should have the same effect

Reply:
Update() returns if the engineModule is null, and that's the only place where the engine module is used, so NREs aren't a problem.